### ISSUE TITLE
How to handle program args

### ISSUE THREAD

**Issue by tubsandcans** [kardianos/service#132]
TLDR; How can I access custom command line flags from within the service program's .Start() method?

I would like to be able to setup the program{} struct with custom fields based on command line flags. This way I can capture and maintain user-provided args during the entire execution of the program.

I'm trying something like 
```
type program struct {
    customFlag string
}
```
and set customFlag to the corresponding os.flags value. This way I can pass those values to one of the exported functions in my go program from within the `Start()` method. For some reason, when I refer to the program instance's customFlag attribute, it's empty even though it was set before passing it to `service.New()`.

**Comment by kardianos** [comment]
os.Args ?

**Comment by tubsandcans** [comment]
os.Args does not work for what I'm trying to do. Here is how I'm setting this up:

```
type Config struct {
	Args []string
}

type program struct {
	exit    chan struct{}
	exec    string
	service service.Service
	*Config
}

//Start wraps the program's run() method
func (p *program) Start(s service.Service) (err error) {
	logger.Error(p.Config)
	p.exit = make(chan struct{})
	go p.run()
	return nil
}

func main() {
	cert := flag.String("cert", "", "[CERT_FILE (.crt | .pem)] set this value to 'skip' if you'd like to skip all verification of certs (INSECURE!)")
	flag.Parse()

	config := &Config{Args: []string{"cert", *cert}}
	prg := &program{
		exec: "lcm",

		Config: config,
	}
	//svcConfig - config for running/installing program as a service
	var svcConfig = &service.Config{
		Name:        "lcm",
		DisplayName: "Log Collection Manager",
		Description: "Log collection management for ReactLive",
		Executable:  lcm.ExecFile,
	}

	s, err := service.New(prg, svcConfig)

        if len(*svcFlag) != 0 {
	        logger.Error("CERT IS ", *cert)
	        serviceCmd(s, *svcFlag)
        }
}
```

When I run this with `service -start -cert skip` I get:
CERT IS skip
printing p.Config in Start() results in `&{[cert ]}`

What is happening to the config struct such that it no longer has *cert?

**Comment by SteelPhase** [comment]
@tubsandcans It appears that your example is incomplete. I don't see `svcFlag` or `serviceCmd` defined. If you'd like your service to start with certain arguments, you'll need to make sure you fill the `Arguments` field in `service.Config`.

**Comment by tubsandcans** [comment]
Yeah I realize I didn't leave it in thinking it was pretty intuitive that it just wraps the service.Control method:
```

func serviceCmd(s service.Service, cmd string) {
	if err := service.Control(s, cmd); err != nil {
		log.Printf("Valid actions: %q\n", service.ControlAction)
		log.Fatal(err)
		logger.Error(err)
	}
}
```
svcFlag is `-service flag`, which I also thought was intuitive. 

Anyways, I tested using Arguments field in service.Config by populating it with an inline declaration of a string array and then printing os.Args in the program. This works, however I cannot get the Arguments field to work with a computed string array, which is useless.

**Comment by kardianos** [comment]
@tubsandcans I didn't fully understand you.

However, you should be aware that []string{"arg1", "arg2"} is not the same as []string{"arg1 arg2}. So if it works with a static []string, then you need to figure out how to get the input into the program.

**Comment by tubsandcans** [comment]
The print statement in the run() method just prints out the running executable filename even though I'm sure I'm setting the Arguments field to a non-empty string array (by printing prgArgs in main())
```
//Start wraps the program's run() method
func (p *program) Start(s service.Service) (err error) {
	p.exit = make(chan struct{})
	go p.run()
	return nil
}

func (p *program) run() (err error) {
                logger.Error(os.Args)
		return
	}
}
var prgArgs []string

func main() {
        svcFlag := flag.String("service", "", "<start|stop|uninstall> - ")
	skipVerify := flag.Bool("skip-verify", false, "Set to skip SSL certificate verification (INSECURE!)")
	ignoreCert := flag.Bool("ignore-cert", false, "Ignore the certificate specified in LCM config for all certificate verification.")
	flag.Parse()

	if *skipVerify {
		prgArgs = append(prgArgs, "-skip-verify")
	}
	if *ignoreCert {
		prgArgs = append(prgArgs, "-no-certificate")
	}

	prg := &program{
		exec: "lcm",

		Config: config,
	}
	//svcConfig - config for running/installing program as a service
	var svcConfig = &service.Config{
		Arguments:   prgArgs,
        }
	s, err := service.New(prg, svcConfig)
	if err != nil {
		log.Fatal(err)
	}

	prg.service = s

	switch {
	case len(*svcFlag) != 0:
		if *svcFlag == "start" {
			logger.Error("in start service command, prgArgs:", prgArgs)
		}
		serviceCmd(s, *svcFlag)
		return
        }
}
```
But If I replace `Arguments: prgArgs` with `Arguments: string[]{"-skip-verify"}`, I can verify it gets passed in to os.Args

**Comment by SteelPhase** [comment]
`s, err := service.New(prg, svcConfig)` should be returning `ErrNameFieldRequired` with the `service.Config` you have provided in your example.

**Comment by kardianos** [comment]
That will only take effect on service install.

**Comment by tubsandcans** [comment]
@SteelPhase: I scrubbed a few things from my original source code. I do have a name for my program, and if I was running into an error like this I wouldn't be wondering why my args aren't getting passed as expected.

**Comment by SteelPhase** [comment]
I don't want to make assumptions about how your code works. It's much better to supply a working example in this case, even if some info has to be changed/redacted.

**Comment by tubsandcans** [comment]
All I wanted to demonstrate were the differences in behavior between setting service.Config's Arguments to a static []string versus setting it to a []string that gets populated with command-line arg values. That is what this issue is about.

**Comment by SteelPhase** [comment]
As @kardianos your arguments will only take affect on service install. There should be no magic difference between a static []string, and a computed []string. If you are having issues with a computed []string you need to look closer at what you are doing to generate it.

- Install service `app -service install -skip-verify`
- Run service `app -service start`

**Comment by SteelPhase** [comment]
@tubsandcans Here's [a working example](https://github.com/SteelPhase/service-test) doing exactly what you want that I just tested.

**Comment by tubsandcans** [comment]
@SteelPhase Thanks for taking the time to put this example together. The problem is once you install a service with a particular set of args, there is no way to change it other than uninstalling and re-installing with the newer args. This is why I was wanting a way to pass in args to a `-service start` command. 

I don't think it's very clear that service.Config's Arguments field only works for `-service install` actions.

**Comment by kardianos** [comment]
You can't do that. It starts automatically. The system starts it.

**Comment by tubsandcans** [comment]
You can manually stop a service, and manually start a service, so yes you CAN do that.

**Comment by SteelPhase** [comment]
Services are usually designed to start at some specific time during boot of a server. I'm not sure this library is designed around on demand services that wouldn't start automatically, but I haven't attempted to use that feature. With that said, manually stopping & starting a service only does just that. With this library if you want to reconfigure the service, you need to uninstall & reinstall it.

A stop, uninstall, install, start process in this library, is the same as a stop, modify the service config file, and start process that you may manually do when managing a service on a linux box yourself.

I think you need to take a step back and figure out what your trying to accomplish and think about how to work with this library to do that. We've laid out to you how to accomplish what you want, within the bounds of this library.

**Comment by LouisAldorio** [comment]
can you help me how to use the config file, I have no idea what are you sayinh about config file, it is not documented in the package. I got the same problem with you, static array of string is working fine but computed from arguments which is os.Args is not working and the service keep terminated, can you help me ?

**Comment by tubsandcans** [comment]
> can you help me how to use the config file, I have no idea what are you sayinh about config file, it is not documented in the package. I got the same problem with you, static array of string is working fine but computed from arguments which is os.Args is not working and the service keep terminated, can you help me ?

It's been a while since I've dealt with this, but I believe I worked around this issue by having my program read from a configuration file (probably .json or .yml) and modify that configuration file before starting the service again.
